tests/ui.rs

```diff
@@ -77,32 +89,52 @@ fn miri_config(target: &str, path: &str, mode: Mode, with_dependencies: bool) ->
 
     let mut config = Config {
         target: Some(target.to_owned()),
-        stderr_filters: stderr_filters().into(),
-        stdout_filters: stdout_filters().into(),
-        mode,
         program,
         out_dir: PathBuf::from(std::env::var_os("CARGO_TARGET_DIR").unwrap()).join("miri_ui"),
-        edition: Some("2021".into()), // keep in sync with `./miri run`
         threads: std::env::var("MIRI_TEST_THREADS")
             .ok()
             .map(|threads| NonZero::new(threads.parse().unwrap()).unwrap()),
         ..Config::rustc(path)
     };
 
+    config.comment_defaults.base().exit_status = match mode {
+        Mode::Pass => Some(0_i32),
+        Mode::Fail => Some(1),
+        Mode::RunDep => None,
+        Mode::Panic => Some(101),
+    }
+    .map(Spanned::dummy)
+    .into();
+
+    config.comment_defaults.base().require_annotations =
+        Spanned::dummy(matches!(mode, Mode::Fail)).into();
+
+    config.comment_defaults.base().normalize_stderr =
+        stderr_filters().iter().map(|(m, p)| (m.clone(), p.to_vec())).collect();
+    config.comment_defaults.base().normalize_stdout =
+        stdout_filters().iter().map(|(m, p)| (m.clone(), p.to_vec())).collect();
+
+    // keep in sync with `./miri run`
+    config.comment_defaults.base().add_custom("edition", Edition("2021".into()));
+
     if with_dependencies {
         // Set the `cargo-miri` binary, which we expect to be in the same folder as the `miri` binary.
         // (It's a separate crate, so we don't get an env var from cargo.)
-        config.dependency_builder.program = {
+        let mut program = CommandBuilder::cargo();
+        program.program = {
             let mut prog = miri_path();
             prog.set_file_name(format!("cargo-miri{}", env::consts::EXE_SUFFIX));
             prog
         };
         let builder_args = ["miri", "run"]; // There is no `cargo miri build` so we just use `cargo miri run`.
-        config.dependency_builder.args = builder_args.into_iter().map(Into::into).collect();
-        config.dependencies_crate_manifest_path =
-            Some(Path::new("test_dependencies").join("Cargo.toml"));
+        program.args = builder_args.into_iter().map(Into::into).collect();
+        let crate_manifest_path = Path::new("test_dependencies").join("Cargo.toml");
         // Reset `RUSTFLAGS` to work around <https://github.com/rust-lang/rust/pull/119574#issuecomment-1876878344>.
-        config.dependency_builder.envs.push(("RUSTFLAGS".into(), None));
+        program.envs.push(("RUSTFLAGS".into(), None));
+        config.comment_defaults.base().set_custom(
```

Comment:
Please move this up, directly below `let crate_manifest_path` -- currently the RUSTFLAGS thing just sits in the middle of the dependency building setup and that doesn't seem to make any sense?

Reply:
I just tried to keep the diff minimal. I'll add a separate commit that refactors this

Reply:
Ah that's much nicer, thanks :)